pkg/rbd/nodeserver.go

```diff
@@ -248,21 +296,118 @@ func (ns *NodeServer) NodeUnpublishVolume(ctx context.Context, req *csi.NodeUnpu
 		return nil, status.Error(codes.NotFound, err.Error())
 	}
 	if notMnt {
-		// TODO should consider deleting path instead of returning error,
-		// once all codes become ready for csi 1.0.
-		return nil, status.Error(codes.NotFound, "volume not mounted")
+		if err = os.RemoveAll(targetPath); err != nil {
+			return nil, status.Error(codes.Internal, err.Error())
+		}
+		return &csi.NodeUnpublishVolumeResponse{}, nil
+	}
+
+	// TODO remove this once  we stop supporting legacy volume delete/mount
```

Comment:
Entire legacy volume handling in Unpublish should not be required. We are not dealing with the RBD image at all in this call.

Reply:
why it's not required, what happens for the already mapped device in publish call. don't we need to unmap the rbd device?

Reply:
Node publish creates target path and bind mounts the staging path to it, hence on Unpublish we need to revert the same. This is true for legacy volumes as well.

We need not unmap the rbd device or even have to detect if it is a legacy volume in the Unpublish (I think the path is independent of the naming, but there could be a corner case).

The unmap should happen in the Unstage, so the detection and calling of ns.umount in line 339 need not be done in the Unpubish call.

Reply:
> Node publish creates target path and bind mounts the staging path to it, hence on Unpublish we need to revert the same. This is true for legacy volumes as well.

how this is true for volumes mounted using v1.0.0  cephcsi? in v1.0.0 targetPath was directly mounted to the device not for the stagingPath right?

> We need not unmap the rbd device or even have to detect if it is a legacy volume in the Unpublish (I think the path is independent of the naming, but there could be a corner case).
> 
> The unmap should happen in the Unstage, so the detection and calling of ns.umount in line 339 need not be done in the Unpubish call.

if the volume and created and mounted using the v1.0.0 (targetpath -> rbddevice)  how to unmount stagingpath from a device is possible?

Reply:
> > Node publish creates target path and bind mounts the staging path to it, hence on Unpublish we need to revert the same. This is true for legacy volumes as well.
> 
> how this is true for volumes mounted using v1.0.0 cephcsi? in v1.0.0 targetPath was directly mounted to the device not for the stagingPath right?
> 
> > We need not unmap the rbd device or even have to detect if it is a legacy volume in the Unpublish (I think the path is independent of the naming, but there could be a corner case).
> > The unmap should happen in the Unstage, so the detection and calling of ns.umount in line 339 need not be done in the Unpubish call.
> 
> if the volume and created and mounted using the v1.0.0 (targetpath -> rbddevice) how to unmount stagingpath from a device is possible?

The above questions are applicable to v1.1.0 volumes as well, if we are dealing with plugin pod upgrades post this change and there are calls to Unstage or Unpublish already mapped/mounted devices by the older plugin version. IOW, upgrade scenarios. My call on upgrade scenarios is that for this change nodes need to be drained and updated (at least for now, till there are other reasons provided).

Once this plugin is rolled out, even for Legacy volumes we would map in Stage and hence would/should unmap in Unstage. Similarly for legacy volumes Publish and Unpublish would not be any different than for non-legacy volumes.

Reply:
have we documented steps for upgrade scenarios?

why the drain is required?  this code handles both legacy and other new volumes without a drain.

Reply:
node drain with the daemonset upgrade  (With OnDelete update strategy (default one) seems correct idea. updating the code